docs/guide/config.md
# 基本配置


| 属性               | 别名        | 类型       | 含义                      |
| ----------------- | ----------- | ------ | ------------------------ |
| width          | w               |  number         | 视频宽度                              |
| height         | h                |    number        | 视频高度                            |
| fps          | rfps                  |  number (默认30)         | 视频帧频    |
| outputDir      | dir                 | string            | 视频输出路径          |
| cacheDir          | 无              |     string    | 缓存目录                        |
| transparent          | 无              |     boolean (默认false)    | 是否启用透明背景（用于生成透明通道视频） |
| output          | out           | string       | 输出文件名(非必须, 在FFCenter模式下不需要填写)                     |
| parallel          | frames    | number （默认5）      | 并行渲染                     |
| highWaterMark          | size     | string (默认1mb)               | node.js [highWaterMark](http://nodejs.cn/api/stream/buffering.html)                          |
| clarity          | renderClarity   | string (模式medium)                  | `medium high low` 画面渲染质量                          |
| cover          | 无           | string         | 封面图                         |
| pool          | 无            |  boolean (默认false)      |  是否启用对象池技术缓存(节省内存模式)                     |
| debug          | 无           |  boolean (默认false)    | boolean Debug模式                    |
| preload          | 无         |     boolean (默认true)      | 是否启用预加载                    |
| antialias  | 无  | boolean | 是否开启平滑 |
| audioLoop  | 无  | boolean （默认true） | 背景音乐是否循环 |
| render  | 无  | string (默认gl) | 使用gl/canvas内核渲染 - 内核cairo/mesa3d |
| ffmpeglog  | 无  | boolean (默认false) | 开启ffmpeg的log |
| vb  | 无  | string | ffmpeg vb |
| crf  | 无  | number (默认crf) | ffmpeg crf |
| preset  | 无  | string | ffmpeg preset |
| cacheFormat  | 无  | string (默认raw) | 缓存图片格式 |
| cacheQuality  | 无  | number (默认80) | 缓存图片质量（非raw格式） |
| defaultOutputOptions  | 无  | boolean | 是否使用默认配置 |

## Useage


```javascript
const creator = new FFCreator({
  cacheDir,                 // 缓存目录
  outputDir,                // 输出目录
  output,                   // 输出文件名(FFCreatorCenter中可以不设)
  width: 500,               // 影片宽
  height: 680,              // 影片高
  cover: 'a.jpg',           // 设置封面
  audioLoop: true,          // 音乐循环
  fps: 24,                  // fps
  threads: 4,               // 多线程(伪造)并行渲染
  debug: false,             // 开启测试模式
  defaultOutputOptions: null,// ffmpeg输出选项配置
});
```


docs/en/guide/config.md
# Basic config


| Attribute | Alias ​​| Type | Meaning |
| ----------------- | ----------- | ------ | ------------ ------------ |
| width | w | number | video width |
| height | h | number | video height |
| fps | None | number (default 30) | Video frame rate |
| outputDir | dir | string | Video output path |
| cacheDir | None | string | Cache directory |
| transparent | None | boolean (default false) | Whether to enable transparent background (used to generate transparent channel video) |
| output | out | string | Output file name (not required, no need to fill in in FFCenter mode) |
| parallel | frames | number (default 5) | parallel rendering |
| highWaterMark | size | string (default 1mb) | node.js [highWaterMark](http://nodejs.cn/api/stream/buffering.html) |
| clarity | renderClarity | string (mode medium) | `medium high low` image rendering quality |
| cover | none | string | cover image |
| pool | None | boolean (default false) | Whether to enable object pool technology caching (memory saving mode) |
| debug | None | boolean (default false) | boolean Debug mode |
| preload | None | boolean (default true) | Whether to enable preload |
| antialias | None | boolean | Whether to enable smoothing |
| audioLoop | None | boolean (default true) | Whether the background music loops |
| render | none | string (default gl) | use gl/canvas kernel rendering-kernel cairo/mesa3d |
| ffmpeglog | None | boolean (default false) | Enable ffmpeg log |
| vb | None | string | ffmpeg vb |
| crf | None | number (default crf) | ffmpeg crf |
| preset | none | string | ffmpeg preset |
| cacheFormat | None | string (default raw) | Cached image format |
| cacheQuality | None | number (default 80) | Cached image quality (non-raw format) |
| defaultOutputOptions | None | boolean | Whether to use the default configuration |


CHANGELOG.md
## v7.0.1
* Add support for output transparent video.
* Add scene time detection function.
* Add srt file format support.
  
## v6.7.1
* defaultOptions supports adding custom ffmpeg commands.
* Upgrade lottie version to remove initialization getapi.

## v6.6.6
* Modified highWaterMark from 1mb to 512kb.
* Update lottie-node version, support more text styles.

## v6.6.1
* Added FFLottie class to support rendering lottie animation files.
* Add FFLottie's example demo file. https://github.com/drawcall/lottie-node

## v6.5.1
* Added FFExtras class, you can freely expand the graphics.
* Upgrade update function.
* add setBlur method.

## v6.3.6
* Modify the stream rendering parallel to be reduced from 5 to 3 to improve performance.
* Fix the event problem of ffcreatorcenter.
* add FFLogger.error to creator.on('complete'.

## v6.2.1
* Add FFRect component, A solid color rectangle component.
* add FFVideo loop Options.
* Modify some configuration options.

## v6.1.2
* Add FFChart component, support most demos of echarts.js. Data visualization video can be made
* Support the synthesis of dynamic data change charts.
* Fix the problem of setScale initialization.
* Refactor the setFont and registerFont functions.

## v5.5.6
* FFVtuber supports dual types of sequence frame and video.
* Delete the addVtuber method of FFCreator.
* Fix the gap problem of scene switching.

## v5.5.2
* Upgrade FFVtuber, now it supports green screen video vtuber.
* remove DateUtil.toSeconds.
* change DateUtil.toMilliseconds method.

## v5.2.7
* Fix audio fadeIn time bug
* Modify some typescript type files

## v5.2.2
* Add a more complete TS type, and the corresponding detection command test-typings and test cases, and install the corresponding ts-node environment.
* Fix the problem that the animation type added by the addEffect method in example/text.js does not exist in the backInBottom in effects.js.
* In addition, some uncertain types are marked with todo.

## v5.0.3
* Remove the renderFps configuration option, unify rfps and fps.
* Rendering performance increased by 60%, time increased by 1/3.

## v4.5.3
* Fix the scale error when FFVideo uses animation.

## v4.5.1
* Fix the background problem of FFSubtitle when there is no text.
* Add the initScale property, fix the problem of animation reset size.

## v4.3.3
* FFAudio supports the interception and playback of long audio.
* When the ffmpeg command is abnormal, kill the process.

## v4.2.6
* Add FFVideoAlbum Component. Support multiple video stitching into one album.
* Optimize ffmpeg command performance.

## v4.1.6
* FFVideo adds precise startTime and endtime control.
* FFNode adds getDurationTime method.

## v4.1.1
* Optimize the FFGifImage component and increase the rendering speed by 30%.
* Upgrade inkpaint version.
* Refactor FFAudio class.

## v4.0.1
* No other reason, because of a user. Thank him https://github.com/tnfe/FFCreator/issues/87

## v3.6.2

* Upgrade [inkpaint](https://github.com/drawcall/inkpaint) rendering engine, double the rendering speed.
* Set the default rendering method of FFCreator to cairo lib.
* Fix the bug that the size of the FFVideo component is incorrect.
* Fix the issue of emit error bubbling in FFCreatorCenter.

## v3.5.1

* Replace the rendering kernel, https://github.com/drawcall/inkpaint. inkpaint has higher rendering speed and performance, and supports more features.
* Add FFGifImage component that supports gif, which can render gif flexibly.
* The text component supports features such as line wrapping, bolding, and stroke.
* Support sound volume, fade in, fade out and other effects.
* Add animation sisters such as blurIn blurOut.
* Support setting cover picture.

## v3.1.2

* Add ffaudio volume support
* Support codec config for FFVideo
* Fix eslint bug

## v3.0.2

* FFCreator3.0+ uses node Stream for data caching. The original version frees up disk space and further improves the speed.
* Refactor the center class, Split into 3 categories.


## v2.0.8

* Add multiple cache formats to save disk
* Clear garbage objects, save memory expenses
* Partial code refactoring, more cohesive
* Add cache format demo example
* Refactor the center class

## v1.4.5 release

* This is a relatively stable version, optimized for many bugs under windows.
* Add center test
* TaskQueue appends task id by default

## v1.3.10

* Add more config option
* Fix any video output bug
* Fix lodash.forEach bug
* Add setFontByFile method


lib/conf/conf.js
'use strict';

/**
 * Conf - A encapsulated configuration class is used to better manage configuration related
 *
 * ####Example:
 *
 *     const conf = new Conf(conf);
 *     const val = conf.getVal(key);
 *     conf.setVal(key, val);
 *
 * @object
 */
const path = require('path');
const mtempy = require('mtempy');
const Utils = require('../utils/utils');

class Conf {
  constructor(conf = {}) {
    this.conf = conf;

    this.conf.pathId = Utils.genUuid();
    this.copyByDefaultVal(conf, 'crf', 20);
    this.copyByDefaultVal(conf, 'vb', null);
    this.copyByDefaultVal(conf, 'cover', null);
    this.copyByDefaultVal(conf, 'pool', false);
    this.copyByDefaultVal(conf, 'debug', false);
    this.copyByDefaultVal(conf, 'preload', true);
    this.copyByDefaultVal(conf, 'antialias', false);
    this.copyByDefaultVal(conf, 'audioLoop', true);
    this.copyByDefaultVal(conf, 'ffmpeglog', false);
    this.copyByDefaultVal(conf, 'transparent', false);
    this.copyByDefaultVal(conf, 'ext', 'mp4');
    this.copyByDefaultVal(conf, 'preset', 'medium');
    this.copyByDefaultVal(conf, 'cacheFormat', 'raw');
    this.copyByDefaultVal(conf, 'cacheQuality', 80);
    this.copyByDefaultVal(conf, 'resolution', 1);
    this.copyByDefaultVal(conf, 'defaultOutputOptions', {
      merge: true,
      options: [],
    });

    this.copyFromMultipleVal(conf, 'fps', 'rfps', 30);
    this.copyFromMultipleVal(conf, 'width', 'w', 800);
    this.copyFromMultipleVal(conf, 'height', 'h', 450);
    this.copyFromMultipleVal(conf, 'parallel', 'frames', 3);
    this.copyFromMultipleVal(conf, 'render', 'renderer', 'canvas');
    this.copyFromMultipleVal(conf, 'outputDir', 'dir', path.join('./'));
    this.copyFromMultipleVal(conf, 'cacheDir', 'cacheDir', mtempy.directory());
    this.copyFromMultipleVal(conf, 'output', 'out', path.join('./', `${this.conf.pathId}.mp4`));
    this.copyFromMultipleVal(conf, 'highWaterMark', 'size', '512kb');
    this.copyFromMultipleVal(conf, 'normalizeAudio', 'norAudio', false);
    this.copyFromMultipleVal(conf, 'clarity', 'renderClarity', 'medium');
  }

  /**
   * Get the val corresponding to the key
   * @param {string} key - key
   * @return {any} val
   * @public
   */
  getVal(key) {
    if (key === 'detailedCacheDir') return this.getCacheDir();
    return this.conf[key];
  }

  /**
   * Set the val corresponding to the key
   * @param {string} key - key
   * @param {any} val - val
   * @public
   */
  setVal(key, val) {
    this.conf[key] = val;
  }

  /**
   * Get the width and height in the configuration (add separator)
   * @param {string} dot - separator
   * @return {string} 'widthxheight'
   * @public
   */
  getWH(dot = 'x') {
    return this.getVal('width') + dot + this.getVal('height');
  }

  /**
   * Get the cache directory
   * @return {string} path
   * @public
   */
  getCacheDir() {
    let cacheDir = this.getVal('cacheDir');
    let pathId = this.getVal('pathId');
    return path.join(cacheDir, pathId);
  }

  copyByDefaultVal(conf, key, defalutVal) {
    this.conf[key] = conf[key] === undefined ? defalutVal : conf[key];
  }

  /**
   * Guarantee that a key must have value
   * @public
   */
  copyFromMultipleVal(conf, key, otherKey, defalutVal) {
    this.conf[key] = conf[key] || conf[otherKey] || defalutVal;
  }

  /**
   * A fake proxy Conf object
   * @public
   */
  static getFakeConf() {
    return fakeConf;
  }
}

const fakeConf = {
  // eslint-disable-next-line
  getVal(key) {
    return null;
  },
  // eslint-disable-next-line
  setVal(key, val) {
    return null;
  },
};

module.exports = Conf;
